File: app/token.py
"""Token Data Structures"""

from enum import Enum, auto
from dataclasses import dataclass

class TokenType(Enum):
    PAREN_OPEN = auto()
    PAREN_CLOSE = auto()
    CURLY_OPEN = auto()
    CURLY_CLOSE = auto()
    SQUARE_OPEN = auto()
    SQUARE_CLOSE = auto()
    COLON = auto()
    COMMA = auto()
    KEYWORD = auto()
    INTEGER = auto()
    FLOAT = auto()
    STRING = auto()
    OPERATOR = auto()
    IDENTIFIER = auto()
    COMMENT = auto()
    RESERVED = auto()

@dataclass
class Token:
    type: TokenType
    value: str
    start_line: int = 0
    end_line: int = 0
    start_column: int = 0
    end_column: int = 0

    def matches(self, other):
        """Check if another token matches this one"""
        return (
            (other.type == self.type) and
            (other.value == self.value)
        )

    def __repr__(self):
        return (
            "Token({}, '{}', line: {}-{}, column: {}-{})".format(
                self.type.name,
                self.value,
                self.start_line,
                self.end_line,
                self.start_column,
                self.end_column
            )
        )

@dataclass
class TokenizeResult:
    consumed_chars: int
    token: Token

    def __repr__(self):
        return f'TokenizeResult({self.consumed_chars}, {self.token})'

File: java/app/src/main/java/org/theolang/compiler/token/Token.java
package org.theolang.compiler.token;

public class Token {
	public final TokenType type;
	public final String value;
	public int startLine;
	public int endLine;
	public int startColumn;
	public int endColumn;

	public Token(
		TokenType type,
		String value
	) {
		this.type = type;
		this.value = value;
		this.startLine = 0;
		this.endLine = 0;
		this.startColumn = 0;
		this.endColumn = 0;
	}

	public Token(
		TokenType type,
		String value,
		int startLine,
		int endLine,
		int startColumn,
		int endcolumn
	) {
		this.type = type;
		this.value = value;
		this.startLine = startLine;
		this.endLine = endLine;
		this.startColumn = startColumn;
		this.endColumn = endColumn;
	}

	public String toString() {
		return "Token({}, '{}', line: {}-{}, column: {}-{})".formatted(
			type.name(),
			value,
			startLine,
			endLine,
			startColumn,
			endColumn
		);
	}
}
